File: engine/tests/fixtures/comprehensive.py
from collections.abc import Iterable
from typing import TypeAlias, TypeVar

T = TypeVar("T")
NumberList: TypeAlias = list[int]

class BaseSolver:
    version: int = 1

class Solver(BaseSolver):
    def __init__(self, values: NumberList) -> None:
        self.values = values

    def solve(self, limit: int = 3) -> dict[int, int]:
        seen: dict[int, int] = {}
        unique: set[int] = set(self.values)
        prefix = self.values[:limit]
        last = self.values[-1]
        for value in prefix:
            if value > 0:
                seen[value] = self.solve(limit - 1)
        while limit > 0:
            limit -= 1
        return seen

def main() -> dict[int, int]:
    return Solver([1, 2, 3]).solve()

File: engine/tests/fixtures/comprehensive.js
import { readFile as read } from "node:fs";

class Solver extends BaseSolver {
  static version = 1;
  values = [];

  constructor(values) {
    this.values = values;
  }

  solve(limit = 3) {
    const seen = new Map();
    const unique = new Set(this.values);
    const prefix = this.values.slice(0, limit);
    for (const value of prefix) {
      if (value > 0) seen.set(value, this.solve(limit - 1));
    }
    while (limit > 0) limit -= 1;
    return seen;
  }
}

export function main() {
  return new Solver([1, 2, 3]).solve();
}
